File: backend/apps/api/serializers/trips.py
"""
Serializers for trips, packages, and related data.
"""
from rest_framework import serializers
from apps.trips.models import (
    Trip, TripPackage, PackageFlight, PackageHotel,
    ItineraryItem, TripUpdate, TripGuideSection,
    ChecklistItem, EmergencyContact, TripFAQ,
    TripMilestone, TripResource
)
from apps.bookings.models import Booking
from apps.content.models import Dua
from apps.pilgrims.models import PilgrimReadiness


class TripListSerializer(serializers.ModelSerializer):
    """Serializer for trip list."""
    
    packages_count = serializers.SerializerMethodField()
    starting_price_minor_units = serializers.SerializerMethodField()
    starting_price_currency = serializers.SerializerMethodField()
    
    class Meta:
        model = Trip
        fields = [
            'id',
            'code',
            'name',
            'slug',
            'excerpt',
            'seo_title',
            'seo_description',
            'cities',
            'commercial_month_label',
            'status',
            'default_nights',
            'starting_price_minor_units',
            'starting_price_currency',
            'start_date',
            'end_date',
            'cover_image',
            'featured',
            'packages_count',
            'updated_at',
        ]
    
    def get_packages_count(self, obj):
        """Get count of available packages."""
        return obj.packages.filter(visibility='PUBLIC').count()

    def get_starting_price_minor_units(self, obj):
        """Return the lowest truthful public package price."""
        package = obj.packages.filter(visibility='PUBLIC', price_minor_units__isnull=False).order_by('price_minor_units').select_related('currency').first()
        return package.price_minor_units if package else None

    def get_starting_price_currency(self, obj):
        """Return the currency for the lowest truthful public package price."""
        package = obj.packages.filter(visibility='PUBLIC', price_minor_units__isnull=False).order_by('price_minor_units').select_related('currency').first()
        if not package or not package.currency:
            return None
        return package.currency.code


class TripDetailSerializer(serializers.ModelSerializer):
    """Serializer for trip details."""
    
    packages_count = serializers.SerializerMethodField()
    
    class Meta:
        model = Trip
        fields = [
            'id', 'code', 'family_code', 'commercial_month_label', 'name', 'cities',
            'status', 'sales_open_date', 'default_nights',
            'start_date', 'end_date', 'visibility',
            'packages_count', 'updated_at'
        ]
    
    def get_packages_count(self, obj):
        """Get count of available packages."""
        return obj.packages.filter(visibility='PUBLIC').count()


class FlightSerializer(serializers.ModelSerializer):
    """Serializer for package flights."""
    
    class Meta:
        model = PackageFlight
        fields = [
            'id', 'leg', 'carrier', 'flight_no',
            'dep_airport', 'dep_dt', 'arr_airport', 'arr_dt',
            'group_pnr'
        ]


class HotelSerializer(serializers.ModelSerializer):
    """Serializer for package hotels."""
    
    class Meta:
        model = PackageHotel
        fields = [
            'id', 'name', 'address', 'room_type',
            'check_in', 'check_out', 'group_confirmation_no'
        ]


class TripPackageSerializer(serializers.ModelSerializer):
    """Serializer for trip package."""
    
    flights = FlightSerializer(many=True, read_only=True)
    hotels = HotelSerializer(many=True, read_only=True)
    currency = serializers.CharField(source='currency.code', read_only=True)
    start_date = serializers.DateField(source='effective_start_date', read_only=True)
    end_date = serializers.DateField(source='effective_end_date', read_only=True)
    nights = serializers.IntegerField(source='effective_nights', read_only=True)
    
    class Meta:
        model = TripPackage
        fields = [
            'id', 'package_code', 'name', 'start_date', 'end_date', 'nights',
            'price_minor_units', 'currency', 'capacity', 'sales_target',
            'hotel_booking_month', 'airline_booking_month', 'status',
            'flights', 'hotels'
        ]


class ItineraryItemSerializer(serializers.ModelSerializer):
    """Serializer for itinerary items."""
    
    attach_url_signed = serializers.SerializerMethodField()
    
    class Meta:
        model = ItineraryItem
        fields = [
            'id', 'day_index', 'start_time', 'end_time',
            'title', 'location', 'notes', 'attach_url_signed', 'updated_at'
        ]
    
    def get_attach_url_signed(self, obj):
        """Get signed URL for attachment."""
        if not obj.attach_public_id:
            return None
        
        from apps.common.cloudinary import signed_delivery
        return signed_delivery(obj.attach_public_id, expires_in=600)


class TripUpdateSerializer(serializers.ModelSerializer):
    """Serializer for trip updates."""
    
    trip_id = serializers.CharField(source='trip.id', read_only=True)
    package_name = serializers.CharField(source='package.name', read_only=True, allow_null=True)
    attach_url_signed = serializers.SerializerMethodField()
    
    class Meta:
        model = TripUpdate
        fields = [
            'id', 'trip_id', 'title', 'body_md', 'urgency', 'pinned',
            'publish_at', 'package_name', 'attach_url_signed',
            'created_at', 'updated_at'
        ]
    
    def get_attach_url_signed(self, obj):
        """Get signed URL for attachment."""
        if not obj.attach_public_id:
            return None
        
        from apps.common.cloudinary import signed_delivery
        return signed_delivery(obj.attach_public_id, expires_in=600)


class GuideSectionSerializer(serializers.ModelSerializer):
    """Serializer for guide sections."""
    
    attach_url_signed = serializers.SerializerMethodField()
    
    class Meta:
        model = TripGuideSection
        fields = ['id', 'order', 'title', 'content_md', 'attach_url_signed']
    
    def get_attach_url_signed(self, obj):
        """Get signed URL for attachment."""
        if not obj.attach_public_id:
            return None
        
        from apps.common.cloudinary import signed_delivery
        return signed_delivery(obj.attach_public_id, expires_in=600)


class ChecklistItemSerializer(serializers.ModelSerializer):
    """Serializer for checklist items."""
    
    package_name = serializers.CharField(source='package.name', read_only=True, allow_null=True)
    
    class Meta:
        model = ChecklistItem
        fields = [
            'id', 'label', 'category', 'is_required',
            'link_url', 'package_name'
        ]


class TripMilestoneSerializer(serializers.ModelSerializer):
    """Serializer for public or pilgrim-visible milestones."""

    package_name = serializers.CharField(source='package.name', read_only=True, allow_null=True)

    class Meta:
        model = TripMilestone
        fields = [
            'id', 'milestone_type', 'title', 'status',
            'target_date', 'actual_date', 'notes', 'package_name', 'updated_at'
        ]


class TripResourceSerializer(serializers.ModelSerializer):
    """Serializer for pilgrim-facing trip resources."""

    package_name = serializers.CharField(source='package.name', read_only=True, allow_null=True)
    file_url_signed = serializers.SerializerMethodField()

    class Meta:
        model = TripResource
        fields = [
            'id', 'title', 'description', 'resource_type', 'viewer_mode',
            'is_pinned', 'published_at', 'file_format', 'metadata',
            'package_name', 'file_url_signed', 'updated_at'
        ]

    def get_file_url_signed(self, obj):
        """Get signed URL for the resource file."""
        if not obj.file_public_id:
            return None

        from apps.common.cloudinary import signed_delivery
        return signed_delivery(obj.file_public_id, expires_in=600)


class PilgrimTripReadinessSerializer(serializers.ModelSerializer):
    """Serializer for a pilgrim's travel-readiness state."""

    booking_reference = serializers.CharField(source='booking.reference_number', read_only=True)
    package_name = serializers.CharField(source='package.name', read_only=True)
    trip_code = serializers.CharField(source='trip.code', read_only=True)
    payment_target_percent = serializers.IntegerField(default=PilgrimReadiness.PAYMENT_TARGET_PERCENT, read_only=True)
    checks = serializers.SerializerMethodField()
    missing_items = serializers.SerializerMethodField()
    blockers = serializers.SerializerMethodField()

    class Meta:
        model = PilgrimReadiness
        fields = [
            'id',
            'booking_reference',
            'trip_code',
            'package_name',
            'status',
            'ready_for_travel',
            'payment_progress_percent',
            'payment_target_percent',
            'validated_at',
            'checks',
            'missing_items',
            'blockers',
            'updated_at',
        ]

    def get_checks(self, obj):
        """Return the component checks that make up the readiness pass."""
        return {
            'profile_complete': obj.profile_complete,
            'passport_valid': obj.passport_valid,
            'visa_verified': obj.visa_verified,
            'documents_complete': obj.documents_complete,
            'payment_target_met': obj.payment_target_met,
            'ticket_issued': obj.ticket_issued,
            'darasa_one_completed': obj.darasa_one_completed,
            'darasa_two_completed': obj.darasa_two_completed,
            'send_off_completed': obj.send_off_completed,
        }

    def get_missing_items(self, obj):
        """Return the outstanding requirements for the pilgrim."""
        return obj.get_missing_requirements()

    def get_blockers(self, obj):
        """Return explicit blockers that need staff intervention."""
        return obj.get_blockers()


class EmergencyContactSerializer(serializers.ModelSerializer):
    """Serializer for emergency contacts."""
    
    class Meta:
        model = EmergencyContact
        fields = ['id', 'label', 'phone', 'hours', 'notes']


class TripFAQSerializer(serializers.ModelSerializer):
    """Serializer for trip FAQs."""
    
    class Meta:
        model = TripFAQ
        fields = ['id', 'question', 'answer', 'order']


class TripEssentialsSerializer(serializers.Serializer):
    """Composite serializer for trip essentials."""
    
    sections = GuideSectionSerializer(many=True)
    checklist = ChecklistItemSerializer(many=True)
    contacts = EmergencyContactSerializer(many=True)
    faqs = TripFAQSerializer(many=True)
    milestones = TripMilestoneSerializer(many=True)
    resources = TripResourceSerializer(many=True)


class DuaSerializer(serializers.ModelSerializer):
    """Serializer for duas."""
    
    class Meta:
        model = Dua
        fields = [
            'id', 'category', 'text_ar', 'text_en',
            'transliteration', 'source'
        ]


class PublicTripDetailSerializer(serializers.ModelSerializer):
    """Comprehensive serializer for public trip details."""
    
    packages = serializers.SerializerMethodField()
    itinerary = serializers.SerializerMethodField()
    faqs = TripFAQSerializer(many=True, read_only=True)
    guide_sections = GuideSectionSerializer(many=True, read_only=True)
    emergency_contacts = EmergencyContactSerializer(many=True, read_only=True)
    has_itinerary = serializers.SerializerMethodField()
    milestones = serializers.SerializerMethodField()
    
    class Meta:
        model = Trip
        fields = [
            'id', 'code', 'family_code', 'commercial_month_label', 'name', 'slug', 'excerpt', 'seo_title', 'seo_description',
            'cities', 'status', 'default_nights', 'start_date', 'end_date',
            'cover_image', 'featured', 'packages', 'itinerary',
            'has_itinerary', 'faqs', 'guide_sections', 'emergency_contacts', 'milestones'
        ]
    
    def get_packages(self, obj):
        """Get public packages with their details."""
        packages = obj.packages.filter(visibility='PUBLIC')
        return TripPackageSerializer(packages, many=True).data
    
    def get_itinerary(self, obj):
        """Get itinerary items."""
        items = obj.itinerary_items.all().order_by('day_index', 'start_time')
        return ItineraryItemSerializer(items, many=True).data
    
    def get_has_itinerary(self, obj):
        """Check if trip has itinerary."""
        return obj.itinerary_items.exists()

    def get_milestones(self, obj):
        """Return only public milestones that can support truthful proof on the website."""
        milestones = obj.milestones.filter(is_public=True).order_by('target_date', 'order', 'created_at')
        return TripMilestoneSerializer(milestones, many=True).data


# ==================== Booking Serializers ====================

class PilgrimBookingCreateSerializer(serializers.ModelSerializer):
    """Serializer for pilgrims to create bookings."""
    
    class Meta:
        model = Booking
        fields = ['id', 'package', 'special_needs']
        read_only_fields = ['id']
    
    def validate_package(self, value):
        """Validate package exists and is public."""
        if value.visibility != 'PUBLIC':
            raise serializers.ValidationError("This package is not available for booking.")
        
        # Check if package has capacity and is full
        if value.capacity:
            booked_count = Booking.objects.filter(
                package=value,
                status__in=['EOI', 'BOOKED', 'CONFIRMED']
            ).count()
            
            if booked_count >= value.capacity:
                raise serializers.ValidationError("This package is fully booked.")
        
        return value
    
    def create(self, validated_data):
        """Create booking with EOI status for authenticated user."""
        request = self.context.get('request')
        user = request.user
        
        # Get or create pilgrim profile
        from apps.accounts.models import PilgrimProfile
        pilgrim = PilgrimProfile.objects.filter(user=user).first()
        
        if not pilgrim:
            if user.role == 'STAFF':
                raise serializers.ValidationError(
                    "Staff users need a pilgrim profile to create bookings. "
                    "Please complete your profile or use the admin dashboard to create bookings."
                )
            else:
                raise serializers.ValidationError("You must complete your profile before booking.")
        
        # Check for duplicate booking (same pilgrim, same package, not cancelled)
        existing = Booking.objects.filter(
            pilgrim=pilgrim,
            package=validated_data['package'],
        ).exclude(status='CANCELLED').first()
        
        if existing:
            raise serializers.ValidationError(
                f"You already have a booking for this package (Reference: {existing.reference_number})."
            )
        
        # Create booking with EOI status
        booking = Booking.objects.create(
            pilgrim=pilgrim,
            status='EOI',  # Expression of Interest
            **validated_data
        )
        
        return booking


class PilgrimBookingDetailSerializer(serializers.ModelSerializer):
    """Serializer for viewing booking details by pilgrims."""
    
    package_id = serializers.CharField(source='package.id', read_only=True)
    trip_id = serializers.CharField(source='package.trip.id', read_only=True)
    trip_name = serializers.CharField(source='package.trip.name', read_only=True)
    trip_code = serializers.CharField(source='package.trip.code', read_only=True)
    trip_date = serializers.DateField(source='package.trip.start_date', read_only=True)
    package_name = serializers.CharField(source='package.name', read_only=True)
    package_price = serializers.IntegerField(source='package.price_minor_units', read_only=True)
    currency_code = serializers.CharField(source='currency.code', read_only=True)
    balance_due = serializers.SerializerMethodField()
    
    class Meta:
        model = Booking
        fields = [
            'id', 'reference_number', 'package_id', 'trip_id', 'trip_name', 'trip_code', 'trip_date',
            'package_name', 'package_price', 'currency_code', 'status',
            'payment_status', 'amount_paid_minor_units', 'balance_due',
            'special_needs', 'ticket_number', 'room_assignment',
            'created_at', 'updated_at'
        ]
    
    def get_balance_due(self, obj):
        """Calculate balance due."""
        package_price = obj.package.price_minor_units or 0
        amount_paid = obj.amount_paid_minor_units or 0
        return max(0, package_price - amount_paid)

File: mobile/lib/api/services/trips.ts
import { apiClient, type ApiResponse } from '../client';
import { API_ENDPOINTS } from '../config';

// Type Definitions for Trip APIs
export interface TripPackage {
  id: string;
  package_code: string;
  name: string;
  start_date: string;
  end_date: string;
  nights: number;
  price_minor_units: number;
  currency: string;
  capacity: number;
  sales_target: number | null;
  hotel_booking_month: string;
  airline_booking_month: string;
  status: string;
  flights: PackageFlight[];
  hotels: PackageHotel[];
}

export interface PackageFlight {
  id: string;
  leg: string;
  carrier: string;
  flight_no: string;
  dep_airport: string;
  dep_dt: string;
  arr_airport: string;
  arr_dt: string;
  group_pnr: string | null;
}

export interface PackageHotel {
  id: string;
  name: string;
  address: string | null;
  room_type: string | null;
  check_in: string;
  check_out: string;
  group_confirmation_no: string | null;
}

export interface ItineraryItem {
  id: string;
  day_index: number;
  start_time: string | null;
  end_time: string | null;
  title: string;
  location: string | null;
  notes: string | null;
  attach_url_signed: string | null;
  updated_at: string;
}

export interface TripFAQ {
  id: string;
  question: string;
  answer: string;
  order: number;
}

export interface TripGuideSection {
  id: string;
  order: number;
  title: string;
  content_md: string;
  attach_url_signed: string | null;
}

export interface EmergencyContact {
  id: string;
  label: string;
  phone: string;
  hours: string | null;
  notes: string | null;
}

export interface Trip {
  id: string;
  code: string;
  commercial_month_label: string | null;
  status: string;
  default_nights: number | null;
  name: string;
  cities: string[];
  start_date: string;
  end_date: string;
  cover_image: string | null;
  featured: boolean;
  packages_count: number;
  updated_at: string;
}

export interface TripDetail {
  id: string;
  code: string;
  family_code?: string;
  commercial_month_label?: string | null;
  status?: string | null;
  sales_open_date?: string | null;
  default_nights?: number | null;
  name: string;
  cities: string[];
  start_date: string;
  end_date: string;
  cover_image: string | null;
  featured: boolean;
  packages: TripPackage[];
  itinerary: ItineraryItem[];
  has_itinerary: boolean;
  faqs: TripFAQ[];
  guide_sections: TripGuideSection[];
  emergency_contacts: EmergencyContact[];
  updated_at?: string;
}

export interface PublicTrip {
  id: string;
  code: string;
  commercial_month_label: string | null;
  status: string;
  default_nights: number | null;
  name: string;
  slug: string;
  excerpt: string;
  seo_title: string;
  seo_description: string;
  cities: string[];
  start_date: string;
  end_date: string;
  cover_image: string | null;
  featured: boolean;
  packages_count: number;
  starting_price_minor_units: number | null;
  starting_price_currency: string | null;
  updated_at: string;
}

export interface PublicTripDetail extends PublicTrip {
  family_code?: string;
  sales_open_date?: string | null;
  packages: TripPackage[];
  itinerary: ItineraryItem[];
  has_itinerary: boolean;
  faqs: TripFAQ[];
  guide_sections: TripGuideSection[];
  emergency_contacts: EmergencyContact[];
  milestones: TripMilestone[];
}

export interface TripMilestone {
  id: string;
  milestone_type: string;
  title: string;
  status: string;
  target_date: string | null;
  actual_date: string | null;
  notes: string;
  package_name: string | null;
  updated_at: string;
}

export interface TripResource {
  id: string;
  title: string;
  description: string;
  resource_type: string;
  viewer_mode: string;
  is_pinned: boolean;
  published_at: string | null;
  file_format: string | null;
  metadata: Record<string, unknown>;
  package_name: string | null;
  file_url_signed: string | null;
  updated_at: string;
}

export interface TripReadinessChecks {
  profile_complete: boolean;
  passport_valid: boolean;
  visa_verified: boolean;
  documents_complete: boolean;
  payment_target_met: boolean;
  ticket_issued: boolean;
  darasa_one_completed: boolean;
  darasa_two_completed: boolean;
  send_off_completed: boolean;
}

export interface TripReadiness {
  id: string;
  booking_reference: string;
  trip_code: string;
  package_name: string;
  status: string;
  ready_for_travel: boolean;
  payment_progress_percent: number;
  payment_target_percent: number;
  validated_at: string | null;
  checks: TripReadinessChecks;
  missing_items: string[];
  blockers: string[];
  updated_at: string;
}

export interface TripEssentials {
  sections: TripGuideSection[];
  checklist: any[];
  contacts: EmergencyContact[];
  faqs: TripFAQ[];
  milestones: TripMilestone[];
  resources: TripResource[];
}

export interface PaginatedResponse<T> {
  count: number;
  next: string | null;
  previous: string | null;
  results: T[];
}

function stringValue(value: unknown): string {
  return typeof value === 'string' ? value : '';
}

function nullableStringValue(value: unknown): string | null {
  return typeof value === 'string' ? value : null;
}

function numberValue(value: unknown): number | null {
  return typeof value === 'number' && !Number.isNaN(value) ? value : null;
}

function arrayValue(value: unknown): unknown[] {
  return Array.isArray(value) ? value : [];
}

function recordValue(value: unknown): Record<string, unknown> {
  return typeof value === 'object' && value !== null ? (value as Record<string, unknown>) : {};
}

function stringArrayValue(value: unknown): string[] {
  return Array.isArray(value) ? value.filter((item): item is string => typeof item === 'string') : [];
}

function computeNights(startDate?: string, endDate?: string) {
  if (!startDate || !endDate) {
    return null;
  }

  const start = new Date(startDate);
  const end = new Date(endDate);
  if (Number.isNaN(start.getTime()) || Number.isNaN(end.getTime())) {
    return null;
  }

  return Math.max(Math.round((end.getTime() - start.getTime()) / (24 * 60 * 60 * 1000)), 0);
}

function normalizePackageFlight(rawValue: unknown): PackageFlight {
  const raw = recordValue(rawValue);
  return {
    id: String(raw.id ?? ''),
    leg: stringValue(raw.leg),
    carrier: stringValue(raw.carrier),
    flight_no: stringValue(raw.flight_no ?? raw.flightNo),
    dep_airport: stringValue(raw.dep_airport ?? raw.depAirport),
    dep_dt: stringValue(raw.dep_dt ?? raw.depDt),
    arr_airport: stringValue(raw.arr_airport ?? raw.arrAirport),
    arr_dt: stringValue(raw.arr_dt ?? raw.arrDt),
    group_pnr: nullableStringValue(raw.group_pnr ?? raw.groupPnr),
  };
}

function normalizePackageHotel(rawValue: unknown): PackageHotel {
  const raw = recordValue(rawValue);
  return {
    id: String(raw.id ?? ''),
    name: stringValue(raw.name),
    address: nullableStringValue(raw.address),
    room_type: nullableStringValue(raw.room_type ?? raw.roomType),
    check_in: stringValue(raw.check_in ?? raw.checkIn),
    check_out: stringValue(raw.check_out ?? raw.checkOut),
    group_confirmation_no: nullableStringValue(raw.group_confirmation_no ?? raw.groupConfirmationNo),
  };
}

function normalizeTripPackage(rawValue: unknown): TripPackage {
  const raw = recordValue(rawValue);
  const startDate = stringValue(raw.start_date ?? raw.startDate);
  const endDate = stringValue(raw.end_date ?? raw.endDate);
  return {
    id: String(raw.id ?? ''),
    package_code: stringValue(raw.package_code ?? raw.packageCode),
    name: stringValue(raw.name),
    start_date: startDate,
    end_date: endDate,
    nights: numberValue(raw.nights) ?? computeNights(startDate, endDate) ?? 0,
    price_minor_units: numberValue(raw.price_minor_units ?? raw.priceMinorUnits) ?? 0,
    currency: stringValue(raw.currency) || 'UGX',
    capacity: numberValue(raw.capacity) ?? 0,
    sales_target: numberValue(raw.sales_target ?? raw.salesTarget),
    hotel_booking_month:
      stringValue(raw.hotel_booking_month ?? raw.hotelBookingMonth),
    airline_booking_month:
      stringValue(raw.airline_booking_month ?? raw.airlineBookingMonth),
    status: stringValue(raw.status),
    flights: arrayValue(raw.flights).map(normalizePackageFlight),
    hotels: arrayValue(raw.hotels).map(normalizePackageHotel),
  };
}

export function normalizePublicTrip(rawValue: unknown): PublicTrip {
  const raw = recordValue(rawValue);
  return {
    id: String(raw.id ?? ''),
    code: stringValue(raw.code),
    commercial_month_label:
      nullableStringValue(raw.commercial_month_label ?? raw.commercialMonthLabel),
    status: stringValue(raw.status),
    default_nights: numberValue(raw.default_nights ?? raw.defaultNights),
    name: stringValue(raw.name),
    slug: stringValue(raw.slug),
    excerpt: stringValue(raw.excerpt),
    seo_title: stringValue(raw.seo_title ?? raw.seoTitle),
    seo_description: stringValue(raw.seo_description ?? raw.seoDescription),
    cities: stringArrayValue(raw.cities),
    start_date: stringValue(raw.start_date ?? raw.startDate),
    end_date: stringValue(raw.end_date ?? raw.endDate),
    cover_image: nullableStringValue(raw.cover_image ?? raw.coverImage),
    featured: Boolean(raw.featured),
    packages_count: Number(raw.packages_count ?? raw.packagesCount ?? 0),
    starting_price_minor_units:
      numberValue(raw.starting_price_minor_units ?? raw.startingPriceMinorUnits),
    starting_price_currency:
      nullableStringValue(raw.starting_price_currency ?? raw.startingPriceCurrency),
    updated_at: stringValue(raw.updated_at ?? raw.updatedAt),
  };
}

export function normalizePublicTripDetail(rawValue: unknown): PublicTripDetail {
  const raw = recordValue(rawValue);
  const packages = arrayValue(raw.packages).map(normalizeTripPackage);
  const pricedPackages = packages.filter((item) => typeof item.price_minor_units === 'number');
  const lowestPrice = pricedPackages
    .slice()
    .sort((left, right) => left.price_minor_units - right.price_minor_units)[0];

  const base = normalizePublicTrip({
    ...raw,
    packages_count: raw.packages_count ?? packages.length,
    starting_price_minor_units:
      raw.starting_price_minor_units ?? lowestPrice?.price_minor_units ?? null,
    starting_price_currency:
      raw.starting_price_currency ?? lowestPrice?.currency ?? null,
    default_nights:
      raw.default_nights ??
      packages.find((item) => typeof item.nights === 'number')?.nights ??
      computeNights(
        stringValue(raw.start_date ?? raw.startDate),
        stringValue(raw.end_date ?? raw.endDate)
      ),
  });

  return {
    ...base,
    family_code: stringValue(raw.family_code ?? raw.familyCode) || undefined,
    sales_open_date: nullableStringValue(raw.sales_open_date ?? raw.salesOpenDate),
    packages,
    itinerary: arrayValue(raw.itinerary).map((item) => {
      const record = recordValue(item);
      return {
        id: String(record.id ?? ''),
        day_index: Number(record.day_index ?? record.dayIndex ?? 0),
        start_time: nullableStringValue(record.start_time ?? record.startTime),
        end_time: nullableStringValue(record.end_time ?? record.endTime),
        title: stringValue(record.title),
        location: nullableStringValue(record.location),
        notes: nullableStringValue(record.notes),
        attach_url_signed: nullableStringValue(record.attach_url_signed ?? record.attachUrlSigned),
        updated_at: stringValue(record.updated_at ?? record.updatedAt),
      };
    }),
    has_itinerary: Boolean(raw.has_itinerary ?? raw.hasItinerary ?? arrayValue(raw.itinerary).length),
    faqs: arrayValue(raw.faqs).map((item) => {
      const record = recordValue(item);
      return {
        id: String(record.id ?? ''),
        question: stringValue(record.question),
        answer: stringValue(record.answer),
        order: Number(record.order ?? 0),
      };
    }),
    guide_sections: arrayValue(raw.guide_sections ?? raw.guideSections)
      .map((item) => {
        const record = recordValue(item);
        return {
          id: String(record.id ?? ''),
          order: Number(record.order ?? 0),
          title: stringValue(record.title),
          content_md: stringValue(record.content_md ?? record.contentMd),
          attach_url_signed: nullableStringValue(record.attach_url_signed ?? record.attachUrlSigned),
        };
      })
      .sort((left, right) => left.order - right.order),
    emergency_contacts: arrayValue(raw.emergency_contacts ?? raw.emergencyContacts).map((item) => {
      const record = recordValue(item);
      return {
        id: String(record.id ?? ''),
        label: stringValue(record.label),
        phone: stringValue(record.phone),
        hours: nullableStringValue(record.hours),
        notes: nullableStringValue(record.notes),
      };
    }),
    milestones: arrayValue(raw.milestones).map((item) => {
      const record = recordValue(item);
      return {
        id: String(record.id ?? ''),
        milestone_type: stringValue(record.milestone_type ?? record.milestoneType),
        title: stringValue(record.title),
        status: stringValue(record.status),
        target_date: nullableStringValue(record.target_date ?? record.targetDate),
        actual_date: nullableStringValue(record.actual_date ?? record.actualDate),
        notes: stringValue(record.notes),
        package_name: nullableStringValue(record.package_name ?? record.packageName),
        updated_at: stringValue(record.updated_at ?? record.updatedAt),
      };
    }),
  };
}

/**
 * Trips Service - handles all trip-related API calls
 */
export class TripsService {
  /**
   * Fetch all public trips (no auth required)
   */
  static async getPublicTrips(featured?: boolean): Promise<ApiResponse<PaginatedResponse<PublicTrip>>> {
    const params = featured !== undefined ? { featured: String(featured) } : undefined;
    const response = await apiClient.get<unknown>(API_ENDPOINTS.TRIPS.PUBLIC_LIST, params);
    if (!response.success || !response.data) {
      return response as ApiResponse<PaginatedResponse<PublicTrip>>;
    }

    const raw = recordValue(response.data);
    const results = arrayValue(raw.results).map(normalizePublicTrip).filter((item) => item.id || item.slug);
    return {
      ...response,
      data: {
        count: Number(raw.count ?? results.length),
        next: nullableStringValue(raw.next),
        previous: nullableStringValue(raw.previous),
        results,
      },
    };
  }

  /**
   * Fetch a single public trip detail (no auth required)
   */
  static async getPublicTripDetail(tripIdentifier: string): Promise<ApiResponse<PublicTripDetail>> {
    const slugResponse = await apiClient.get<unknown>(API_ENDPOINTS.TRIPS.PUBLIC_DETAIL_BY_SLUG(tripIdentifier));
    if (slugResponse.success && slugResponse.data) {
      return {
        ...slugResponse,
        data: normalizePublicTripDetail(slugResponse.data),
      };
    }

    const idResponse = await apiClient.get<unknown>(API_ENDPOINTS.TRIPS.PUBLIC_DETAIL(tripIdentifier));
    if (idResponse.success && idResponse.data) {
      return {
        ...idResponse,
        data: normalizePublicTripDetail(idResponse.data),
      };
    }

    return {
      success: false,
      error: idResponse.error || slugResponse.error || 'Unable to load this journey right now.',
      status: idResponse.status || slugResponse.status,
    };
  }

  /**
   * Fetch user's trips (requires auth)
   */
  static async getMyTrips(token: string): Promise<ApiResponse<Trip[]>> {
    return apiClient.get<Trip[]>(API_ENDPOINTS.TRIPS.MY_TRIPS, undefined, token);
  }

  /**
   * Fetch a single trip detail for authenticated user
   */
  static async getMyTripDetail(tripId: string, token: string): Promise<ApiResponse<TripDetail>> {
    return apiClient.get<TripDetail>(API_ENDPOINTS.TRIPS.MY_TRIP_DETAIL(tripId), undefined, token);
  }

  /**
   * Fetch itinerary for a trip (requires auth)
   */
  static async getTripItinerary(tripId: string, token: string): Promise<ApiResponse<ItineraryItem[]>> {
    return apiClient.get<ItineraryItem[]>(API_ENDPOINTS.TRIPS.ITINERARY(tripId), undefined, token);
  }

  /**
   * Fetch trip essentials (guide sections, checklist, contacts, FAQs)
   */
  static async getTripEssentials(tripId: string, token: string): Promise<ApiResponse<{
    sections: TripGuideSection[];
    checklist: any[];
    contacts: EmergencyContact[];
    faqs: TripFAQ[];
    milestones: TripMilestone[];
    resources: TripResource[];
  }>> {
    return apiClient.get<TripEssentials>(API_ENDPOINTS.TRIPS.ESSENTIALS(tripId), undefined, token);
  }

  /**
   * Fetch public milestones for a booked trip
   */
  static async getTripMilestones(tripId: string, token: string): Promise<ApiResponse<TripMilestone[]>> {
    return apiClient.get<TripMilestone[]>(API_ENDPOINTS.TRIPS.MILESTONES(tripId), undefined, token);
  }

  /**
   * Fetch published trip resources for a booked trip
   */
  static async getTripResources(tripId: string, token: string): Promise<ApiResponse<TripResource[]>> {
    return apiClient.get<TripResource[]>(API_ENDPOINTS.TRIPS.RESOURCES(tripId), undefined, token);
  }

  /**
   * Fetch travel-readiness status for a booked trip
   */
  static async getTripReadiness(tripId: string, token: string): Promise<ApiResponse<TripReadiness>> {
    return apiClient.get<TripReadiness>(API_ENDPOINTS.TRIPS.READINESS(tripId), undefined, token);
  }
}
